lib/purescript-ast/src/Language/PureScript/Environment.hs

```diff
@@ -655,3 +627,17 @@ dictSynonymName = ProperName . dictSynonymName' . runProperName
 
 isDictSynonym :: ProperName a -> Bool
 isDictSynonym = T.isSuffixOf "$Dict" . runProperName
+
+-- |
+-- Generate a list of nominal roles, whose length is the same as the number of
+-- type arguments of types of the given kind.
+--
+nominalRolesForKind :: Type a -> [Role]
+nominalRolesForKind
+  = go []
+  where
+    go acc = \case
+      TypeApp _ (TypeApp _ fn k1) _k2 | eqType fn tyFunction ->
+        go (Nominal : acc) k1
+      _k ->
+        Nominal : acc
```

Comment:
We should recurse on the second argument of `->` and return the accumulator untouched in the base case. Otherwise we return `[Nominal]` for `Type`, `[Nominal, Nominal]` for `Type -> Type` and `Type -> Type -> … -> Type`, `[Nominal, Nominal, Nominal]` for `(Type -> Type) -> Type`…

```purs
module Example where

import Safe.Coerce (coerce)

foreign import data Foreign3 :: Type -> Type -> Type -> Type

-- We infer `nominal nominal` instead of `nominal nominal nominal`,
-- this is equivalent to giving a phantom role to `Foreign3` last parameter:

example :: ∀ a b c d. Foreign3 a b c -> Foreign3 a b d
example = coerce
```

Also now that we have kind polymorphism we should recurse under `ForAll`:

```purs
module Example where

import Safe.Coerce (coerce)

foreign import data Foreign2 :: ∀ k. k -> k -> Type

-- We infer `nominal` instead of `nominal nominal`,
-- this is equivalent to giving a phantom role to `Foreign2` last parameter:

example :: ∀ k (a :: k) (b :: k) (c :: k). Foreign2 a b -> Foreign2 a c
example = coerce
```

I understand this function was just renamed and extracted from `Language.PureScript.TypeChecker.Roles` so I can open an issue and deal with this in another pull request if you so wish.

Reply:
Dealing with this in another pull request sounds preferable to me. Thanks!